src/cargo/core/resolver/dep_cache.rs

```diff
@@ -265,6 +265,8 @@ pub fn resolve_features<'b>(
     let reqs = build_requirements(s, opts)?;
     let mut ret = Vec::new();
     let mut used_features = HashSet::new();
+    // Add `default` as a feature that must exist.
+    used_features.insert("default".into());
```

Comment:
I think the only read of `used_features` is below at `.filter(|s| !used_features.contains(s))`.
`Std::HashSet` only allocates when something is added, and this set is only used if there are enabled optional dependencies. So maybe if we add a `.filter(|s| s != "default")` then we can avoid the allocation here.
(yes, I know, this is micro optimizing when not on the hot path.)